packages/components/react/src/Panels/WorkspacePanel.tsx

```diff
@@ -21,41 +17,80 @@ interface Props {
  * This component is the orchestrator between various interactive components.
  */
 export function WorkspacePanel({ tutorialStore, theme }: Props) {
-  const fileTree = tutorialStore.hasFileTree();
   const hasEditor = tutorialStore.hasEditor();
   const hasPreviews = tutorialStore.hasPreviews();
   const hideTerminalPanel = !tutorialStore.hasTerminalPanel();
 
-  const editorPanelRef = useRef<ImperativePanelHandle>(null);
-  const previewPanelRef = useRef<ImperativePanelHandle>(null);
   const terminalPanelRef = useRef<ImperativePanelHandle>(null);
-  const previewRef = useRef<ImperativePreviewHandle>(null);
   const terminalExpanded = useRef(false);
 
-  const [helpAction, setHelpAction] = useState<'solve' | 'reset'>('reset');
+  return (
+    <PanelGroup className={resizePanelStyles.PanelGroup} direction="vertical">
+      <Editor
+        theme={theme}
+        tutorialStore={tutorialStore}
+        hasEditor={hasEditor}
+        hasPreviews={hasPreviews}
+        hideTerminalPanel={hideTerminalPanel}
+      />
 
-  const selectedFile = useStore(tutorialStore.selectedFile);
-  const currentDocument = useStore(tutorialStore.currentDocument);
+      <PanelResizeHandle
+        className={resizePanelStyles.PanelResizeHandle}
+        hitAreaMargins={{ fine: 5, coarse: 5 }}
+        disabled={!hasEditor}
+      />
 
-  const lesson = tutorialStore.lesson!;
+      <Preview
+        theme={theme}
+        tutorialStore={tutorialStore}
+        terminalPanelRef={terminalPanelRef}
+        terminalExpanded={terminalExpanded}
+        hideTerminalPanel={hideTerminalPanel}
+        hasPreviews={hasPreviews}
+        hasEditor={hasEditor}
+      />
 
-  const onEditorChange = useCallback<OnEditorChange>((update) => {
-    tutorialStore.setCurrentDocumentContent(update.content);
-  }, []);
+      <PanelResizeHandle
+        className={resizePanelStyles.PanelResizeHandle}
+        hitAreaMargins={{ fine: 5, coarse: 5 }}
+        disabled={hideTerminalPanel || !hasPreviews}
+      />
 
-  const onEditorScroll = useCallback<OnEditorScroll>((position) => {
-    tutorialStore.setCurrentDocumentScrollPosition(position);
-  }, []);
+      <Terminal
+        tutorialStore={tutorialStore}
+        theme={theme}
+        terminalPanelRef={terminalPanelRef}
+        terminalExpanded={terminalExpanded}
+        hideTerminalPanel={hideTerminalPanel}
+        hasEditor={hasEditor}
+        hasPreviews={hasPreviews}
+      />
+    </PanelGroup>
+  );
+}
 
-  const onFileSelect = useCallback((filePath: string | undefined) => {
-    tutorialStore.setSelectedFile(filePath);
-  }, []);
+function Editor({ theme, tutorialStore, hasEditor }: PanelProps) {
+  const [helpAction, setHelpAction] = useState<'solve' | 'reset'>('reset');
+  const selectedFile = useStore(tutorialStore.selectedFile);
+  const currentDocument = useStore(tutorialStore.currentDocument);
+  const lesson = tutorialStore.lesson!;
 
-  const onHelpClick = useCallback(() => {
+  function onHelpClick() {
```

Comment:
It's unnecessary as `EditorPanel` and its `FileTab` are not using `React.memo`. They will re-render when ever their parent `WorkspacePanel` or `Editor` re-renders.

We should only use `useCallback` in cases where it's explicitly required. Adding them unnecessarily may lead to unintended complexity. What do you think?

Reply:
> We should only use `useCallback` in cases where it's explicitly required. Adding them unnecessarily may lead to unintended complexity. What do you think?

What unintended complexity are you referring to? Do you mean thinking about dependencies?

I tend to think that component should provide stable references to components they render regardless of whether those direct components use memo or not. Especially because those component might pass that props down their tree, and down the line there will likely be a memoized component that now is never memoized. Which leads to unexpected perf issues.

Additionally, I like that it makes explicit what a callback's dependencies are and sometimes you realised that they can be tweaked to have none (like when using a `setState` to access a value as opposed to accessing the state directly).

When `useCallback` isn't used then downstream components can't just benefit from adding `memo` and it becomes a lot more annoying to inspect every usage up the rendering tree.

If by default we use `useCallback` everywhere in component higher up in the tree then it's easier to use `memo` and improve perf when needed.

Curious to hear @d3lm thoughts on the topic.

Reply:
Is there a performance issue that we are solving here, or just optimizing (which isn't actually working at the moment due to lack of `React.memo`) for future development?

> What unintended complexity are you referring to?

Unnecessary use of `useCallback` and `useMemo`. When they are used without `React.memo`, they don't provide any help. In React codebases you should not default to `useCallback` - use it only when it's required. I'm quite sure there was good explanation for this in https://overreacted.io/ when React `16.8` was released.

> downstream components can't just benefit from adding `memo` and it becomes a lot more annoying to inspect every usage up the rendering tree.

Yep, the `CodeMirrorEditor` here is doing exactly that. It's not expecting the callback props to be stable, so it's creating refs for them.

> If by default we use `useCallback` everywhere in component higher up in the tree then it's easier to use `memo` and improve perf when needed.

If we start doing this, should we also start to add other performance optimizations as well? No unstable default function arguments, `memo` and `useMemo` everywhere? Or just add https://react.dev/learn/react-compiler. These should be opt-in when needed.